core/primitives/src/block_header.rs

```diff
@@ -592,6 +593,155 @@ impl BlockHeader {
+        )
+    }
+
+    /// Creates a new BlockHeader from information in the view of an existing block.  
+    pub fn from_view(
+        expected_hash: &CryptoHash,
+        epoch_protocol_version: ProtocolVersion,
+        height: BlockHeight,
+        prev_hash: CryptoHash,
+        block_body_hash: CryptoHash,
+        prev_state_root: MerkleHash,
+        prev_chunk_outgoing_receipts_root: MerkleHash,
+        chunk_headers_root: MerkleHash,
+        chunk_tx_root: MerkleHash,
+        outcome_root: MerkleHash,
+        timestamp: u64,
+        challenges_root: MerkleHash,
+        random_value: CryptoHash,
+        prev_validator_proposals: Vec<ValidatorStake>,
+        chunk_mask: Vec<bool>,
+        block_ordinal: NumBlocks,
+        epoch_id: EpochId,
+        next_epoch_id: EpochId,
+        next_gas_price: Balance,
+        total_supply: Balance,
+        challenges_result: ChallengesResult,
+        signature: Signature,
+        last_final_block: CryptoHash,
+        last_ds_final_block: CryptoHash,
+        epoch_sync_data_hash: Option<CryptoHash>,
+        approvals: Vec<Option<Box<Signature>>>,
+        next_bp_hash: CryptoHash,
+        block_merkle_root: CryptoHash,
+        prev_height: BlockHeight,
+        chunk_endorsements: Option<ChunkEndorsementsBitmap>,
+    ) -> Self {
+        let header = Self::new_impl(
+            epoch_protocol_version,
+            epoch_protocol_version,
+            epoch_protocol_version,
+            height,
+            prev_hash,
+            block_body_hash,
+            prev_state_root,
+            prev_chunk_outgoing_receipts_root,
+            chunk_headers_root,
+            chunk_tx_root,
+            outcome_root,
+            timestamp,
+            challenges_root,
+            random_value,
+            prev_validator_proposals,
+            chunk_mask,
+            block_ordinal,
+            epoch_id,
+            next_epoch_id,
+            next_gas_price,
+            total_supply,
+            challenges_result,
+            None, // signer
+            Some(signature),
+            last_final_block,
+            last_ds_final_block,
+            epoch_sync_data_hash,
+            approvals,
+            next_bp_hash,
+            block_merkle_root,
+            prev_height,
+            chunk_endorsements,
+        );
+        // Note: We do not panic but only log if the hash of the created header does not match the expected hash (From the view)
+        // because there are tests that check if we can downgrade a BlockHeader's view a previous version, in which case the hash
+        // of the header changes.
+        if header.hash() != expected_hash {
+            tracing::debug!(height, header_hash=?header.hash(), ?expected_hash, "Hash of the created header does not match expected hash");
+        }
+        header
+    }
+
+    /// Common logic for generating BlockHeader for different purposes, including new blocks, from views, and for genesis block
+    fn new_impl(
```

Comment:
We could probably unite arguments in BlockHeaderInnerLite, BlockHeaderInnerRestV1 (?) and the rest  earlier to reduce number of arguments.

Reply:
It would be great to reduce the number of arguments (it is way too many now), but not sure how to do that at this point, let me submit this and I will reconsider that in a new PR.